File: custom_alnahran/custom_alnahran/utilis/js/sales_invoice.js
frappe.ui.form.on("Sales Invoice Item" , {
  item_code: function (frm  , cdt , cdn){
    let row = locals[cdt][cdn]

    if(row.item_code){

      if(frm.doc.set_warehouse){
        frappe.call({
          method :"custom_alnahran.custom_alnahran.utilis.py.sales_invoice.get_last_purchaseand_stock_rate",  
          args:{
            "item_code":  row.item_code,
            "warehouse": frm.doc.set_warehouse      
          } ,
          callback : function (r){
            frappe.model.set_value(cdt, cdn, 'custom_last_purchase_rate', r.message[0])
            frappe.model.set_value(cdt, cdn, 'custom_last_stock_transfer_rate', r.message[1])
          }
        })
      }
      else{
        frappe.show_alert({ message: "Source Warehouse not selected to fetch last purchase and stock transfer rate.", indicator: "green" });
      }
    }
  }
})

File: custom_alnahran/custom_alnahran/utilis/py/sales_invoice.py
import frappe 

@frappe.whitelist()
def get_last_purchaseand_stock_rate(item_code, warehouse):
    purchase_entries = frappe.db.get_all("Purchase Invoice Item",
                                      filters={'item_code': item_code, 'docstatus':1, 'parenttype':"Purchase Invoice", "warehouse": warehouse},
                                      fields=['rate'],
                                      order_by='creation desc',
                                      limit=1)
                                      
    stock_entries = frappe.db.get_all("Stock Entry Detail",
                                      filters={'item_code': item_code, 'docstatus':1, "t_warehouse": warehouse},
                                      fields=['basic_rate'],
                                      order_by='creation desc',
                                      limit=1)

    return (purchase_entries[0].rate if purchase_entries else 0) , (stock_entries[0].basic_rate if stock_entries else 0)
